### PR TITLE
perf(windows): cut per-pane memory and cold start on the startup path

### PR DESCRIPTION
**Pull request by amanthanvi**
## Summary

Five fixes to the Windows startup, IO and renderer paths. Every one was found by
the #121 benchmark baseline rather than by inspection.

Split out of #191 on review feedback. A first attempt at that split left the
call sites behind in #191, so this PR defined the machinery but nothing invoked
it; that is fixed — the spawn-site configuration, the
`beginRendererFrameUpdate` hook and the `prepareSurfaceDeinit` call now live
here with the code they activate. #191 stacks on this branch.

## The changes

**Windows thread stacks.** `internal_os.surfaceThreadSpawnConfig()` gives the
per-surface renderer, IO and PTY-reader threads a 1 MiB stack argument. Zig
passes `stack_size` to `CreateThread` without
`STACK_SIZE_PARAM_IS_A_RESERVATION`, so this sets the initial **commit**; the
reservation still comes from the PE header (16 MiB) and threads still grow on
demand. Only the upfront commit charge changes, so no new stack-overflow
ceiling is introduced.

**ConPTY read batch.** The Windows PTY reader used a 16 KiB buffer, making the
reader the bottleneck on bursty output. Now 64 KiB, with no runtime override —
one product path, which is also the one the benchmark measures.

**Idle wake loop.** The renderer thread armed a 500 ms cursor-blink timer
unconditionally. It now arms only when the terminal actually requests a blinking
cursor, so a steady DECSCUSR cursor lets the process go genuinely idle.

**Repaint reservation.** The renderer reserves the repaint slot before the
expensive `updateFrame` work so a paint completing during that work cannot be
lost, and `completeRendererRepaintCoalesce` closes the interleaving where paint
completion clears the reservation and checks `retry_pending` between the
renderer observing a pending reservation and publishing its own retry.

Review raised a reservation-leak freeze on the early returns in
`drawFrameWithReservation`. Every early exit now releases a reservation it did
not hand to a paint. On the reachability of the original report: `hasVsync()` is
a compile-time constant `false` in this fork (`DisplayLink = void`, and the
Win32 branch is excluded), and `renderOnce` checks the same visibility and vsync
conditions immediately before reserving with no event-loop turn in between, so
neither early return could actually be reached holding a reservation, and there
is no cross-thread read of `vsync_enabled` to synchronize. The fix is still
worth having — the invariant currently depends on two functions agreeing — but
it is hardening, not a live freeze. That is recorded in a comment at the site.

**Direct default framebuffer.** Win32 surfaces render straight to the default
framebuffer when a runtime GL query proves the back buffer is sRGB-encoded and
blending is linear, with no custom shaders. Everything else keeps the owned
offscreen target and the strategy fails closed. Proving that precondition needs
real pixel-format selection, so WGL selection lives in
`src/apprt/win32/pixel_format.zig`.

Also in this PR:

- `printCell` cleared the wide-tail marker only under `runtime_safety`, so a
  ReleaseFast build could hand `clearCells` a row whose wide head and tail
  disagreed. Now unconditional, with a regression test. This is a terminal-core
  semantic change in release builds and is called out here deliberately; it was
  found by running terminal workloads under ReleaseFast for the benchmark.
- Renderer teardown rebinds the owning WGL context before deleting graphics
  objects (`prepareSurfaceDeinit`).
- Font discovery prefetch overlaps Windows font enumeration with runtime startup.

## Measured effect — stale, and not attributable to this PR alone

An earlier revision of this body presented a before/after table as this PR's
"measured effect". That was wrong twice over and has been removed: the run
predates the split and reorganization so the binary no longer exists, and at the
time of that run the changes were not partitioned the way they are now, so no
column could be attributed to this branch by itself.

The numbers that motivated the work, for context only: per-pane memory 104 MB ->
32.14 MB, cold start 484 ms -> 297 ms median, idle successful swaps 26/10s -> 0,
measured 2026-08-27 against `d031dc4` with all of this and the instrumentation in
one build. Treat them as the reason these fixes exist, not as this PR's result.
A fresh interactive run on the merged stack is required before any of it is
quoted as a current claim; `docs/windows-benchmark-methodology.md` in #191 says
the same.

## Validation

- `zig build -Demit-exe=true` — pass
- `zig build -Demit-exe=true -Dtarget=aarch64-windows-msvc -Dcpu=baseline` — pass
  (the previous revision was x64-only and broke ARM64; every `wglGetProcAddress`
  cast now has the `@alignCast` that aarch64's stricter fn-pointer alignment
  requires)
- `zig build test -Demit-test-exe=true` — pass, 0 failures
- `zig fmt --check` on every touched file, `scripts/check-source-format.ps1` — pass

## Bot-review triage (2026-08-30)

One review finding, real, fixed: `changeConfig` set `reinitialize_shaders` for
a custom-shader change but did not bump `target_config_modified`, so a
`reload_config` that enabled a custom shader without also resizing never re-ran
target selection. On a Win32 surface holding a direct fbo-0 target, the newly
created custom shader would have rendered straight to the default framebuffer,
bypassing the no-custom-shader precondition the direct-default strategy is
gated on; the reverse transition stranded an offscreen target until the next
resize. Both inputs to `initTarget` now go through one named predicate,
`targetConfigChangeRequiresRecreate`, with a unit test pinning the invariant.

Two independent reviews reported this same defect.

## Trims

`pixel_format.zig` lost 128 more lines on review: the dead
`wgl_bootstrap_cleanup_order`/`WglBootstrapCleanupStep` pair, the test-only
`validateWglMultisampleSelection` and `wglChooseAttributeValue`, the tautological
`pixelFormatSelectionSource` asserts and the helper itself, the always-zero
`set_pixel_format_count` provenance field, and the three-state `WglRestoreOutcome`
enum whose `.restored` arm was unreachable (both call sites hardcoded `false`).
With the 156 lines of rejection-counter diagnostics cut earlier that is 284 lines
out of this file.

Not taken: merging `betterWglPixelFormatCandidate` and
`betterClassicPixelFormatCandidate` into one generic comparator. It is a
reasonable ~90-line saving, but it rewrites the ranking that decides which pixel
format the renderer gets, and I cannot re-validate that without an interactive
desktop run. Flagging rather than doing it blind.

## Summary by Sourcery

Improve Windows startup, rendering, and terminal I/O efficiency while hardening framebuffer selection, repaint coordination, and teardown correctness.

New Features:
- Render eligible Windows surfaces directly to the default sRGB framebuffer when configuration and runtime capabilities permit.
- Select and validate Windows WGL pixel formats with verified sRGB support and safe fallback behavior.

Bug Fixes:
- Prevent lost or permanently coalesced repaints during renderer updates and ensure repaint reservations are released on early exits.
- Avoid unnecessary cursor-blink wakeups for steady cursors and improve renderer follow-up scheduling.
- Preserve terminal page integrity when overwriting the tail of a wide character in release builds.
- Recreate render targets when custom-shader configuration changes so framebuffer strategy remains valid.
- Restore the owning graphics context before renderer resource teardown.

Enhancements:
- Reduce Windows per-surface thread stack commitment and increase ConPTY read batches.
- Overlap Windows font discovery with application startup.

Tests:
- Add regression and policy tests covering Windows thread stacks, font prefetch, WGL selection, framebuffer strategy, repaint scheduling, cursor blinking, target recreation, terminal integrity, and target ownership.

## Summary by CodeRabbit

* **New Features**
  * Windows startup now begins font discovery in the background, improving launch responsiveness.
  * OpenGL rendering can use the default framebuffer when appropriate, reducing unnecessary presentation work.
  * Windows graphics initialization now selects and validates pixel formats more reliably.
  * Large Windows terminal output is handled with a larger read buffer.

* **Bug Fixes**
  * Improved repaint scheduling, cursor blinking behavior, renderer teardown, and wide-character editing integrity.
  * Added diagnostics for pixel-format startup failures.

### CODE REVIEW

**Comment by amanthanvi** [comment]
@coderabbitai review
@codex review

Context for reviewers:

Per-pane memory and cold-start work on the Windows startup path: 1 MiB thread stacks at the surface/termio spawn sites, a renderer repaint reservation that closes a WM_PAINT-versus-renderer-thread race, a font-discovery prefetch, and a direct-framebuffer (fbo 0) render path that skips the offscreen blit. This is the **bottom** of a two-PR stack; #191 sits on top of it.

Prior adversarial review already settled two things — please do not re-litigate either:

1. **1 MiB stacks are SAFE.** Zig 0.15.2 passes `stack_size` to `CreateThread` with `dwCreationFlags = 0`, i.e. *without* `STACK_SIZE_PARAM_IS_A_RESERVATION`, so the parameter sets initial **commit** while the **reserve** still comes from the PE header. The reviewer built an exe with this exact toolchain and read its optional header: `SizeOfStackReserve` = 16 MiB, `SizeOfStackCommit` = 4 KiB. Each thread still reserves 16 MiB of address space and grows on demand through guard pages. The 1 MiB figure changes only the upfront commit charge; there is no new stack-overflow ceiling on the deep paths (the 64 KiB `read_buf`, HarfBuzz shaping, oniguruma, xev callbacks were all checked).
2. **The direct-framebuffer sRGB path is colour-correct and fails closed.** The gate is not WGL provenance: `OpenGL.init` queries `GL_FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING` on `GL_BACK_LEFT` of fbo 0 at **runtime** and requires `GL_SRGB`, plus `blending.isLinear()`, plus no custom shaders, plus binding == 0; any query error falls back to the offscreen path with a warning. Old and new paths perform identical decode / blend / encode under the same global `GL_FRAMEBUFFER_SRGB`.

Please scrutinise most:
1. **Reservation lifetime in `renderer/Thread.zig`** — every early return out of `drawFrameWithReservation` must cancel a held reservation, or `renderer_repaint_requested` latches true and the pane freezes permanently. The vsync and visibility early-returns are the ones to check hardest.
2. **`hasVsync()` cross-thread reads** — `vsync_enabled` / `swap_interval_configured` / `swap_interval_supported` are flipped by the app thread inside `ensureWin32SwapInterval` and `changeConfig` while the renderer thread reads them.
3. **`@ptrCast` alignment on every `wglGetProcAddress` result** in `pixel_format.zig` — function-pointer alignment versus `*const anyopaque` differs on aarch64-windows.

File: src/renderer/generic.zig [hunk 1]
```diff
@@ -424,7 +427,11 @@ pub fn Renderer(comptime GraphicsAPI: type) type {
                 if (self.custom_shader_state) |*state| {
                     try state.resize(api, width, height);
                 }
-                const target = try api.initTarget(width, height);
+                const target = try api.initTarget(
+                    width,
+                    height,
+                    self.custom_shader_state != null,
+                );
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/renderer/generic.zig:481]: **REAL — fixed.**

Verified against source. `changeConfig` at `src/renderer/generic.zig:2066-2068` set `reinitialize_shaders` for a custom-shader change and nothing else, while the target is only re-created when the size changed or `target_config_modified` moved (`src/renderer/generic.zig:1585-1596`). `drawFrame` does update `self.has_custom_shaders` first (`generic.zig:1542-1547` -> `initShaders` at `generic.zig:855-865`) and does create/destroy `frame.custom_shader_state` (`generic.zig:1555-1567`), so the shader state changes while the target does not. With a `.default_framebuffer` target from `OpenGL.initTarget` (`src/renderer/OpenGL.zig:462-479`), enabling a custom shader without a resize leaves fbo 0 in place and the shader renders straight to it, defeating the `!opts.custom_shaders` clause in `targetStrategy` (`src/renderer/OpenGL.zig:60-69`). The reverse transition stranded the offscreen target as you describe.

This is the second independent report of the same defect; an earlier manual review flagged the identical missing `target_config_modified` bump.

Fix: both inputs to `initTarget` — blending mode and custom-shader presence — now go through one named predicate, `targetConfigChangeRequiresRecreate` (`src/renderer/generic.zig:41-66`), which `changeConfig` uses to bump `target_config_modified` in either direction. Test `renderer target recreation tracks custom shader presence` (`src/renderer/generic.zig:68-88`) pins it; confirmed it actually executes (`zig build test -Dtest-filter=...` selects exactly one more test than the empty-filter baseline: 68 vs 67).

Gates on the fix: `zig fmt --check src` clean apart from the pre-existing generated `src\build\uucode_tables.zig`; `zig build -Demit-exe=true` exit 0; `zig build test -Demit-test-exe=true` exit 0; `scripts/check-source-format.ps1` exit 0.

**Comment by amanthanvi** [comment]
@coderabbitai review
@codex review

Round 2, pushed as `fc0e05cc2`. Your round-1 finding was real and is fixed.

**Custom-shader toggle left a stale render target.** Confirmed exactly as reported: `changeConfig` set `reinitialize_shaders` and nothing else, while the target is recreated only when the size changes or `target_config_modified` moves. `drawFrame` does update `has_custom_shaders` and does create/destroy `frame.custom_shader_state`, so the shader state changed while the target did not — and with a `.default_framebuffer` target from `OpenGL.initTarget`, enabling a custom shader without a resize left fbo 0 in place and rendered the shader straight to it, bypassing the deliberate no-custom-shader fail-closed gate. The reverse transition stranded an offscreen target and forfeited the optimization until an unrelated resize. Target recreation is now driven by a change in custom-shader *presence* in both directions, with a test pinning the declared invariant.

An independent prior review had found the identical defect, so two sources agreed and it needed no re-derivation.

Please concentrate on:
1. **The recreation trigger.** Is presence-change the right key, or can a config reload change shader *content* without changing presence in a way that also needs a new target? And confirm the recreation is idempotent — it must not thrash a target on every frame if the predicate is ever mis-evaluated.
2. **Teardown ordering** around the swapped target: an old default-framebuffer target must not be destroyed as if it owned an FBO, and an old offscreen target must actually release its texture.
3. **The fail-closed gate itself** — with the fix in place, is there any remaining path where a custom shader can render to fbo 0?

**Settled, please do not re-litigate:** (a) 1 MiB thread stacks are safe — Zig passes `stack_size` to `CreateThread` with `dwCreationFlags = 0`, i.e. without `STACK_SIZE_PARAM_IS_A_RESERVATION`, so it sets initial COMMIT while the PE header still reserves 16 MiB, verified by reading a built exe's optional header (`SizeOfStackReserve` 16 MiB, `SizeOfStackCommit` 4 KiB). (b) The direct-framebuffer sRGB path is colour-correct and fails closed — `OpenGL.init` queries `GL_FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING` on `GL_BACK_LEFT` of fbo 0 at runtime and requires `GL_SRGB`, plus linear blending, no custom shaders and binding 0; any query error falls back to offscreen.

Note this is the bottom of a two-PR stack; #191 sits on top and merges after it.

File: src/apprt/win32/pixel_format.zig [hunk 2]
```diff
@@ -0,0 +1,1422 @@
...
+                break :load error.WglExtensionsUnavailable;
+            };
+            const functions: WglPixelFormatFunctions = functions: {
+                // ARB_framebuffer_sRGB is specified in terms of the EXT
+                // pixel-format API. Keep that family exact when available.
+                if (hasWglExtension(extensions, "WGL_EXT_pixel_format") and
+                    hasWglExtension(extensions, "WGL_ARB_framebuffer_sRGB"))
+                {
+                    const choose_raw = wglGetProcAddress("wglChoosePixelFormatEXT");
+                    const get_raw = wglGetProcAddress("wglGetPixelFormatAttribivEXT");
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/apprt/win32/pixel_format.zig]: **Partly correct — the reachability concern is real and now fixed, but the spec reasoning is inverted.**

## The API-family claim is backwards (declined)

The comment says the ARB framebuffer-sRGB attribute "belongs with `wglChoosePixelFormatARB`/`wglGetPixelFormatAttribivARB` while the EXT procedures belong with `WGL_EXT_framebuffer_sRGB`". The registry says the opposite — *both* framebuffer-sRGB extensions are written against the **EXT** pixel-format API.

`ARB_framebuffer_sRGB.txt`, Dependencies:

> WGL_EXT_extensions_string is required for WGL support.
> WGL_EXT_pixel_format is required for WGL support.

New Tokens:

> Accepted by the `<piAttributes>` parameter of **wglGetPixelFormatAttribivEXT**, wglGetPixelFormatAttribfvEXT, and the `<piAttribIList>` and `<pfAttribIList>` of **wglChoosePixelFormatEXT**:
> `WGL_FRAMEBUFFER_SRGB_CAPABLE_ARB   0x20A9`

`EXT_framebuffer_sRGB.txt` is identical in both sections and declares `WGL_FRAMEBUFFER_SRGB_CAPABLE_EXT` at the *same* value, `0x20A9`. The `_ARB` suffix names the extension that defines the token, not the entry-point family it is valid for.

Conversely `WGL_EXT_colorspace.txt`:

> WGL_ARB_pixel_format is required. This extension is written against WGL_ARB_pixel_format extension.
> Accepted as an attribute name by **wglGetPixelFormatAttribivARB** ... and **wglChoosePixelFormatARB**: `WGL_COLORSPACE_EXT 0x309D`

So both pre-existing branches were already spec-exact: `pixel_format.zig:643-670` paired `WGL_EXT_pixel_format` + `WGL_ARB_framebuffer_sRGB` with the EXT entry points, and `WGL_ARB_pixel_format` + `WGL_EXT_colorspace` with the ARB entry points. Swapping them as suggested would have paired `WGL_COLORSPACE_EXT` with entry points it is not valid for.

## The reachability gap is real (fixed)

The underlying worry stands, for a different reason. Requiring `WGL_EXT_pixel_format` for the only widely-implemented sRGB attribute is not the portable choice: a driver advertising the standard `WGL_ARB_pixel_format` + `WGL_ARB_framebuffer_sRGB` pair without `WGL_EXT_pixel_format` matched neither branch and fell through to `enumerateClassicPixelFormat`, so the verified-sRGB direct-framebuffer strategy silently never engaged. Nothing accepted `WGL_EXT_framebuffer_sRGB` either.

That the ARB pairing is safe in practice is not a guess — it is what mainstream loaders do. GLFW (`src/wgl_context.c`) gates its entire modern path on `WGL_ARB_pixel_format`, never checks `WGL_EXT_pixel_format`, and queries the attribute through the ARB entry point:

```c
if (_glfw.wgl.ARB_framebuffer_sRGB || _glfw.wgl.EXT_framebuffer_sRGB)
    ADD_ATTRIB(WGL_FRAMEBUFFER_SRGB_CAPABLE_ARB);
...
wglGetPixelFormatAttribivARB(window->context.wgl.dc, pixelFormat, 0, attribCount, attribs, values)
```

### What changed

`src/apprt/win32/pixel_format.zig` now carries an explicit, documented pairing table (`wgl_srgb_api_family_order`) tried in order, with each family's entry points and sRGB attribute resolved together and never mixed:

1. `WGL_EXT_pixel_format` + either framebuffer-sRGB spelling → EXT entry points, attribute `0x20A9` (spec-exact; widened to accept `WGL_EXT_framebuffer_sRGB`, which was previously ignored).
2. `WGL_ARB_pixel_format` + `WGL_EXT_colorspace` → ARB entry points, `WGL_COLORSPACE_EXT` (unchanged).
3. **New:** `WGL_ARB_pixel_format` + either framebuffer-sRGB spelling → ARB entry points, attribute `0x20A9`.

The new family is tried **last** deliberately, so this is purely additive: any driver that already matched a spec-exact pairing keeps exactly the path it had, and the new branch only fires where the code previously gave up. `wglMultisampleQuerySupported` was extended to keep the ARB families on `WGL_ARB_multisample` exactly, since `WGL_SAMPLE_BUFFERS_ARB`/`WGL_SAMPLES_ARB` are declared against the ARB entry points.

Your closing note is right and worth restating: this only changes whether the runtime `GL_FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING` query on `GL_BACK_LEFT` is *reached*. That query still gates the strategy and still fails closed to the offscreen path, and a driver that rejects the attribute fails the candidate query and falls back to classic selection exactly as before.

### Tests

Five new tests in `pixel_format.zig`, all driver-free:

- family order covers every pairing exactly once;
- the ARB-only regression: `WGL_ARB_pixel_format WGL_ARB_framebuffer_sRGB` with no `WGL_EXT_pixel_format` now resolves instead of falling back, plus the `WGL_EXT_framebuffer_sRGB`-only spellings;
- spec-exact pairings still win when advertised, and `WGL_EXT_colorspace` keeps priority over the de-facto pairing so no working driver changes path;
- cross-family combinations are refused (notably `WGL_EXT_pixel_format` + `WGL_EXT_colorspace` resolves to nothing), as are pixel-format APIs with no sRGB attribute and sRGB attributes with no pixel-format API;
- each family's sRGB attribute stays bound to its own entry points, and the multisample query pairs with each family's own extension.

**Comment by amanthanvi** [comment]
@codex review

Round 3, now at `7e5c87f34`.

Your round-2 finding was **half wrong and half right**, and the half that was right mattered. Full reasoning is on the thread; briefly:

The spec claim was **inverted**. Both framebuffer-sRGB specs pair attribute `0x20A9` with the **EXT** entry points, not the ARB ones. `ARB_framebuffer_sRGB.txt` says "WGL_EXT_pixel_format is required for WGL support" and lists `WGL_FRAMEBUFFER_SRGB_CAPABLE_ARB 0x20A9` as accepted by `wglGetPixelFormatAttribivEXT` / `wglChoosePixelFormatEXT`; `EXT_framebuffer_sRGB.txt` is identical with the same value. Conversely `WGL_EXT_colorspace.txt` requires `WGL_ARB_pixel_format` and is accepted by `wglGetPixelFormatAttribivARB`. So both pre-existing branches were already spec-exact, and the suggested swap would have paired `WGL_COLORSPACE_EXT` with entry points it is not valid for — which would have disabled the path on drivers where it currently works.

The **reachability gap was real**: a driver exposing `WGL_ARB_pixel_format` + `WGL_ARB_framebuffer_sRGB` but not `WGL_EXT_pixel_format` matched neither branch and silently fell back to classic, and `WGL_EXT_framebuffer_sRGB` was never accepted at all. Fixed with a documented ordered pairing table. The de-facto ARB family (what GLFW's `wgl_context.c` actually does — gate on `WGL_ARB_pixel_format`, never check EXT, query `0x20A9` through `wglGetPixelFormatAttribivARB`) is tried **last**, so the change is purely additive and no already-working driver changes path. Five new driver-free tests.

Please concentrate on:
1. **The pairing table's ordering.** Is "spec-exact families first, de-facto ARB family last" the right precedence, and can any driver now match a family whose entry points it does not actually export?
2. Whether the runtime `GL_FRAMEBUFFER_ATTACHMENT_COLOR_ENCODING` check still gates every path that reaches the direct-framebuffer strategy — that check, not the extension string, is the real safety net.
3. Anything in the new table that could request an attribute a driver rejects, which would disable the optimization rather than merely failing to enable it.

**Settled, do not re-litigate:** 1 MiB thread stacks are safe (Zig passes `stack_size` to `CreateThread` with `dwCreationFlags = 0`, so it sets COMMIT while the PE header reserves 16 MiB — verified from a built exe's optional header), and the direct-framebuffer sRGB path is colour-correct and fails closed.

**Comment by amanthanvi** [comment]
@coderabbitai review

@codex review

Final pre-merge review pass. This PR has already been through adversarial review and at least one bot round; findings were triaged as fix / decline-with-evidence / follow-up rather than applied blindly, so some flagged items are deliberate design with reasoning recorded in-thread. Please focus on anything that would be wrong *at merge time*: stale claims in the description versus the current head, cross-PR interactions, and correctness of security-relevant paths. If you flag something previously declined, please engage with the recorded reasoning rather than restating the original finding.

**Comment by amanthanvi** [comment]
@coderabbitai full review
@greptileai review this draft
@codex review

**Comment by amanthanvi** [comment]
@coderabbitai review

@greptileai

@codex review

File: src/renderer/generic.zig (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/renderer/generic.zig:481]: Addressed on the current head by `fc0e05cc2`: config changes now bump `target_config_modified` whenever custom-shader presence changes, and `renderer target recreation tracks custom shader presence` pins both directions. The emitted full suite passed on the current head.

File: src/Surface.zig [hunk 3]
```diff
@@ -862,6 +862,12 @@ pub fn deinit(self: *Surface) void {
 
         // We need to become the active rendering thread again
         self.renderer.threadEnter(self.rt_surface) catch unreachable;
+        // Renderer resources belong to this runtime surface's graphics
+        // context. Teardown is back on the app thread now, so restore that
+        // context before deleting any graphics objects.
+        self.renderer.prepareSurfaceDeinit(self.rt_surface) catch |err| {
+            log.err("error preparing renderer surface deinit err={}", .{err});
+        };
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/Surface.zig:878]: Fixed in `4f53193ff`. A failed owning-context rebind now suppresses `renderer.deinit()` so no GL deletion can run against a missing or sibling context; the rest of `Surface` teardown still proceeds. The exceptional path deliberately leaves those resources for WGL context destruction rather than risking cross-pane corruption. The emitted full suite passed.

[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/Surface.zig:878]: Fixed in `4f53193ff`. A failed owning-context rebind now suppresses `renderer.deinit()` so no GL deletion can run against a missing or sibling context; the rest of `Surface` teardown still proceeds. The exceptional path deliberately leaves those resources for WGL context destruction rather than risking cross-pane corruption. The emitted full suite passed.

[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/Surface.zig:878]: Fixed in `4f53193ff`. A failed owning-context rebind now suppresses `renderer.deinit()` so no GL deletion can run against a missing or sibling context; the rest of `Surface` teardown still proceeds. The exceptional path deliberately leaves those resources for WGL context destruction rather than risking cross-pane corruption. The emitted full suite passed.

File: src/apprt/win32/pixel_format.zig [hunk 4]
```diff
@@ -0,0 +1,1422 @@
...
+
+fn isWglPixelFormatHardError(err: anyerror) bool {
+    return switch (err) {
+        error.WglBootstrapMakeCurrentFailed,
+        error.WglBootstrapRestoreCurrentFailed,
+        error.WglBootstrapClearCurrentFailed,
+        => true,
+        else => false,
+    };
+}
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/apprt/win32/pixel_format.zig:322]: Fixed in `4f53193ff`. `WglBootstrapMakeCurrentFailed` is no longer a hard pixel-format error: because the failed call leaves the previous pair current and neither bootstrap path has touched the real child DC, selection falls back to the classic path. Restore/clear failures remain hard. Added a focused classification regression test; focused tests, build, and the emitted full suite passed.

File: src/apprt/win32/pixel_format.zig [hunk 5]
```diff
@@ -0,0 +1,1699 @@
...
+        hasWglExtension(extensions, "WGL_EXT_framebuffer_sRGB");
+}
+
+fn wglSrgbApiFamilyAdvertised(
+    family: WglSrgbApiFamily,
+    extensions: []const u8,
+) bool {
+    return switch (family) {
+        .ext_pixel_format_framebuffer_srgb => hasWglExtension(extensions, "WGL_EXT_pixel_format") and
+            hasWglFramebufferSrgbExtension(extensions),
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/apprt/win32/pixel_format.zig:569]: Declined as a spec mismatch. The official Khronos `ARB_framebuffer_sRGB` spec requires `WGL_EXT_pixel_format` and defines `WGL_FRAMEBUFFER_SRGB_CAPABLE_ARB` for `wglGetPixelFormatAttribivEXT` / `wglChoosePixelFormatEXT` (Dependencies and New Tokens): https://registry.khronos.org/OpenGL/extensions/ARB/ARB_framebuffer_sRGB.txt . `EXT_framebuffer_sRGB` is likewise written against the EXT pixel-format API: https://registry.khronos.org/OpenGL/extensions/EXT/EXT_framebuffer_sRGB.txt . Therefore the current EXT-first arm is the spec-exact pairing even when the advertised framebuffer extension name is ARB. The ARB-entry-point arm remains a last-resort de-facto compatibility path only when no spec-exact EXT family is available.

File: src/renderer/Thread.zig [hunk 6]
```diff
@@ -660,18 +689,38 @@ fn renderOnce(self: *Thread, from_wakeup: bool) bool {
...
     }
-    self.renderer.updateFrame(
+
+    // updateFrame returns the cursor mode from the same locked terminal
+    // snapshot used to rebuild this frame.
+    const frame_update: ?rendererpkg.Renderer.FrameUpdate = self.renderer.updateFrame(
         self.state,
         self.flags.cursor_blink_visible,
-    ) catch |err|
+    ) catch |err| frame_update: {
```
[the remark this answers is not in the document]
Comment (amanthanvi) [reply src/renderer/Thread.zig:847]: Addressed in f83da9f86: restored the renderer-thread draw-request, frame-update, async-wakeup, and follow-up trace hooks at their corresponding event boundaries. Targeted render-trace/follow-up tests and the emitted executable build pass. The stacked #191 branch integrates this commit and retains its richer successful-frame state handoff.

**Comment by amanthanvi** [comment]
@coderabbitai review

@greptileai review

@codex review

**Comment by amanthanvi** [comment]
Rebased onto latest main after #184 landed; preserved the power-aware pacing and repaint-reservation semantics.

@coderabbitai review

@greptileai review

@codex review
